File: js/Chess/.solved/ChessProblem_Solved.js
import PieceColor from '../PieceColor'
import PieceType from '../PieceType'
import ChessStatus from '../ChessStatus'

export default class ChessProblem {
    constructor(board) {
        this.board = board;
    }

    getStatusFor(color) {
        const isCheck = this.isCheckFor(color);
        const hasMoves = this.hasMovesFor(color, this.isValidMove);
        if (isCheck){
            return hasMoves ? ChessStatus.check : ChessStatus.mate;
        } else {
            return hasMoves ? ChessStatus.ok : ChessStatus.stalemate;
        }
    }

    isCheckFor(color) {
        return this.hasMovesFor(invertColor(color), (from, to) => {
            const piece = this.board.getPiece(to);
            return piece ? piece.is(color, PieceType.King) : false;
        })
    }

    hasMovesFor(color, isAcceptableMove) {
        const moves = this.board.getPieces(color).filter(pieceLoc =>
            this.getMoves(pieceLoc).filter(destLoc => isAcceptableMove(pieceLoc, destLoc)).length
        );

        return Boolean(moves.length);
    }

    getMoves(pieceLoc) {
        return this.board.getPiece(pieceLoc).getMoves(pieceLoc, this.board)
    }

    isValidMove = (from, to) => {
        const pieceColor = this.board.getPiece(from).color;
        const move = this.board.performTemporaryMove(from, to);
        const isValid = !this.isCheckFor(pieceColor);
        move.undo();
        return isValid;
    }
}

function invertColor(color) {
    return color === PieceColor.black ? PieceColor.white : PieceColor.black;
}

File: cs/clean-code/Chess/Solved/ChessProblem_Solved.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Chess.Solved
{
    public class ChessProblem_Solved
    {
        private readonly Board board;

        public ChessProblem_Solved(Board board)
        {
            this.board = board;
        }

        public ChessStatus GetStatusFor(PieceColor color)
        {
            var isCheck = IsCheckFor(color);
            var hasMoves = HasMovesFor(color, IsValidMove);
            if (isCheck)
                return hasMoves ? ChessStatus.Check : ChessStatus.Mate;
            else
                return hasMoves ? ChessStatus.Ok : ChessStatus.Stalemate;
        }

        private bool IsCheckFor(PieceColor color) =>
            HasMovesFor(color.Invert(),
                (from, to) => Piece.Is(board.GetPiece(to), color, PieceType.King));

        private bool HasMovesFor(PieceColor color,
            Func<Location, Location, bool> isAcceptableMove)
        {
            var moves =
                from pieceLoc in board.GetPieces(color)
                from destLoc in GetMoves(pieceLoc)
                where isAcceptableMove(pieceLoc, destLoc)
                select true;
            return moves.Any();
        }

        private IEnumerable<Location> GetMoves(Location pieceLoc) => 
            board.GetPiece(pieceLoc).GetMoves(pieceLoc, board);

        private bool IsValidMove(Location from, Location to)
        {
            var pieceColor = board.GetPiece(from).Color;
            var move = board.PerformTemporaryMove(from, to);
            var isValid = !IsCheckFor(pieceColor);
            move.Undo();
            return isValid;
        }
    }

    public static class PieceColorExtensions
    {
        public static PieceColor Invert(this PieceColor color) => 
            color == PieceColor.Black ? PieceColor.White : PieceColor.Black;
    }
}